pkg/build/prometheus/metrics_via_collectors.go

```diff
@@ -0,0 +1,150 @@
+	BuildStageHistogram                = "status_stages_duration"
+	BuildStageHistogramBucketQuery     = BuildSubsystem + separator + BuildStageHistogram + separator + "bucket"
+	BuildStageHistogramCountQuery      = BuildSubsystem + separator + BuildStageHistogram + separator + "count"
+	BuildStageHistogramSumQuery        = BuildSubsystem + separator + BuildStageHistogram + separator + "sum"
+	BuildStageStepHistogram            = "status_stages_steps_duration"
+	BuildStageStepHistogramBucketQuery = BuildSubsystem + separator + BuildStageStepHistogram + separator + "bucket"
+	BuildStageStepHistogramCountQuery  = BuildSubsystem + separator + BuildStageStepHistogram + separator + "count"
+	BuildStageStepHistogramSumQuery    = BuildSubsystem + separator + BuildStageStepHistogram + separator + "sum"
+)
+
+var (
+	// decided not to have a separate counter for failed builds, which have reasons,
+	// vs. the other "finished" builds phases, where the reason is not set
+	descBuildsCounter = prometheus.NewDesc(
+		BuildSubsystem+separator+BuildCounter,
+		"Counts total teriminal builds by phase, and in case of failures, reason",
+		[]string{"phase", "reason", "namespace", "name"},
+		nil,
+	)
+	descBuildsGauge = prometheus.NewDesc(
+		BuildSubsystem+separator+BuildGauge,
+		"Show current counts of active builds by phase, with start time in milliseconds",
+		[]string{"phase", "namespace", "name", "startime"},
+		nil,
+	)
+	descBuildsOverallHistogram = prometheus.NewDesc(
+		BuildSubsystem+separator+BuildHistogram,
+		"Histogram of durations in seconds of terminated builds, by termination type",
+		[]string{"phase", "namespace", "name"},
+		nil,
+	)
+	descBuildStagesHistogram = prometheus.NewDesc(
+		BuildSubsystem+separator+BuildStageHistogram,
+		"Histogram of durations in milliseconds of stages that have finished, by stage name",
+		[]string{"stage", "namespace", "name"},
+		nil,
+	)
+	descBuildStageStepsHistogram = prometheus.NewDesc(
+		BuildSubsystem+separator+BuildStageStepHistogram,
+		"Histogram of durations in milliseconds of steps that have finished, by stage name and step name",
+		[]string{"stage", "step", "namespace", "name"},
+		nil,
+	)
+)
+
+type buildCollector struct {
+	lister internalversion.BuildLister
+}
+
+// kube-state-metrics leverages prometheus.Registerer but does not exist in the version of prometheus
+// vendored into origin
+
+func IntializeMetricsCollector(buildLister internalversion.BuildLister) {
+	bc := buildCollector{lister: buildLister}
+	prometheus.MustRegister(&bc)
+}
+
+// Describe implements the prometheus.Collector interface.
+func (bc *buildCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- descBuildsCounter
+	ch <- descBuildsGauge
+	ch <- descBuildsOverallHistogram
+	ch <- descBuildStagesHistogram
+	ch <- descBuildStageStepsHistogram
+}
+
+// Collect implements the prometheus.Collector interface.
+func (bc *buildCollector) Collect(ch chan<- prometheus.Metric) {
```

Comment:
correct, and that is how all these existing k8s implementations (both already in origin's vendor list and the kube-state-metrics) is doing it; they just hit their store/informers right out of the Collect method ... no more sophistication of additional batching/caching layers